### PR TITLE
feat: implement ephemeral partition encryption

### PR DESCRIPTION
**Pull request by Unix4ever**
Fixes: https://github.com/talos-systems/talos/issues/3106

This PR introduces the first part of disk encryption support.
New config section `systemDiskEncryption` was added into MachineConfig.
For now it contains only Ephemeral partition encryption.

Encryption itself supports two kinds of keys for now:
- generated which can be generated from a template.
- static key, which is hardcoded into the config (mainly for test purposes).

Talosctl cluster create can now be told to encrypt ephemeral partition
by using `--encrypt-ephemeral` flag.

Additionally:
- updated pkgs library version.
- changed Dockefile to copy cryptsetup deps from pkgs.

Signed-off-by: Artem Chernyshev <artem.0xD2@gmail.com>

### CODE REVIEW

File: cmd/installer/pkg/install/manifest.go [hunk 1]
```diff
@@ -550,6 +546,10 @@ func (t *Target) Partition(pt *gpt.GPT, pos int, bd *blockdevice.BlockDevice) (e
 		return err
 	}
 
+	if err = pt.Write(); err != nil {
```
Comment (smira) [inline cmd/installer/pkg/install/manifest.go:549]: why are we writing partition table here?

File: cmd/installer/pkg/install/verify.go [hunk 2]
```diff
@@ -66,7 +66,7 @@ func VerifyDiskAvailability(devpath, label string) (err error) {
 
 	part, err := dev.GetPartition(label)
 	if err != nil {
-		return err
+		return nil
```
Comment (smira) [inline cmd/installer/pkg/install/verify.go:69]: this doesn't look right?

File: internal/integration/base/api.go [hunk 3]
```diff
@@ -241,22 +241,27 @@ func (apiSuite *APISuite) WaitForBootDone(ctx context.Context) {
...
+							return
+						}
 					}
 				}
 			}
-		}
-	}, client.WithTailEvents(-1)))
+		}, client.WithTailEvents(-1))
 
+		return retry.ExpectedError(e)
```
Comment (smira) [inline internal/integration/base/api.go:261]: I don't think we should retry _every_ error here, as it probably hides some legitimate errors?

and you said that error was about context deadline exceeded, so it won't help here?

File: internal/pkg/encryption/keys/generated.go [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+package keys
+
+import (
+	"fmt"
+
+	"github.com/google/uuid"
+	"github.com/talos-systems/go-smbios/smbios"
+)
+
+// GeneratedKeyHandler generates the key based on current node information
```
Comment (smira) [inline internal/pkg/encryption/keys/generated.go:14]: I think @Ulexus might have a better idea, but something like `DeterminsticKeyHandler` better describes what is going on here.

File: internal/pkg/encryption/keys/generated.go [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	machineUUID, err := s.SystemInformation().UUID()
+	if err != nil {
+		return nil, err
+	}
+
+	if machineUUID == uuid.Nil {
```
Comment (smira) [inline internal/pkg/encryption/keys/generated.go:41]: we need to check for randomness of UUID here - could be a follow-up ticket, but we need to have it in 0.9

File: internal/pkg/mount/mount.go [hunk 6]
```diff
@@ -257,9 +280,21 @@ func (p *Point) Mount() (err error) {
 // Unmount attempts to retry an unmount on EBUSY. It will attempt a
 // retry every 100 milliseconds over the course of 5 seconds.
 func (p *Point) Unmount() (err error) {
-	p.target = path.Join(p.Prefix, p.target)
-	if err := mountRetry(unmount, p, true); err != nil {
+	var mounted bool
+
+	if mounted, err = p.IsMounted(); err != nil {
 		return err
+	} else if mounted {
```
Comment (smira) [inline internal/pkg/mount/mount.go:287]: nit: we don't need `else` here, as previous block ends with `return`

File: internal/pkg/mount/system.go [hunk 7]
```diff
@@ -75,16 +81,84 @@ func SystemMountPointForLabel(device *blockdevice.BlockDevice, label string, opt
...
+	// Format the partition if it does not have any filesystem
+	preMountHooks = append(preMountHooks, func(p *Point) error {
+		sb, err := filesystem.Probe(p.source)
+		if err != nil {
+			return err
+		}
+
+		p.fstype = ""
+
+		// no filesystem detected should format it
```
Comment (smira) [inline internal/pkg/mount/system.go:143]: comment out of sync with the code?

File: pkg/machinery/config/types/v1alpha1/v1alpha1_types.go [hunk 8]
```diff
@@ -1051,6 +1069,45 @@ type DiskPartition struct {
...
+	//   description: >
+	//     Key slot number for luks2 encryption.
+	KeySlot int `yaml:"slot"`
+}
+
+// EncryptionKeyStatic represents throw away key type.
+type EncryptionKeyStatic struct {
+	//   description: >
+	//     Defines the static key value.
+	KeyData string `yaml:"key,omitempty"`
```
Comment (smira) [inline pkg/machinery/config/types/v1alpha1/v1alpha1_types.go:1105]: if we really need this, I guess it should be base64 encoded to allow for binary keys.

File: pkg/machinery/config/types/v1alpha1/v1alpha1_types.go [hunk 9]
```diff
@@ -1051,6 +1069,45 @@ type DiskPartition struct {
...
+}
+
+// EncryptionKeyGenerated represents deterministically generated key from the node UUID and PartitionLabel.
+type EncryptionKeyGenerated struct{}
```
Comment (smira) [inline pkg/machinery/config/types/v1alpha1/v1alpha1_types.go:1109]: Same as above, probably need some better name

File: cmd/installer/pkg/install/install.go [hunk 10]
```diff
@@ -116,37 +116,35 @@ func (i *Installer) probeBootPartition() error {
 		dev, err := blockdevice.Open(i.options.Disk)
 		if err != nil {
 			i.bootPartitionFound = false
-
-			return err
```
[the remark this answers is not in the document]
Comment (Unix4ever) [reply cmd/installer/pkg/install/install.go:120]: Yeah, I think missed it somehow. Thanks.

File: cmd/installer/pkg/install/install.go [hunk 11]
```diff
@@ -183,7 +181,7 @@ func (i *Installer) Install(seq runtime.Sequence) (err error) {
 
 	// Mount the partitions.
 
-	mountpoints, err := i.manifest.SystemMountpoints()
+	mountpoints, err := i.manifest.SystemMountpoints(mount.WithFlags(mount.SkipIfNoFilesystem | mount.SkipIfEncrypted))
```
[the remark this answers is not in the document]
Comment (Unix4ever) [reply cmd/installer/pkg/install/install.go:184]: :heavy_check_mark: yeah

File: cmd/installer/pkg/install/manifest.go [hunk 12]
```diff
@@ -379,6 +377,10 @@ func (m *Manifest) preserveContents(device Device, targets []*Target) (err error
 	anyPreserveContents := false
 
 	for _, target := range targets {
+		if target.Label == constants.EphemeralPartitionLabel && target.PreserveContents {
```
[the remark this answers is not in the document]
Comment (Unix4ever) [reply cmd/installer/pkg/install/manifest.go:380]: sure, that'll be more readable

File: internal/integration/base/api.go (hunk 3 above)
Comment (Unix4ever) [reply internal/integration/base/api.go:261]: yeah, that's just my experiments. I will revert that back

File: internal/pkg/encryption/keys/generated.go (hunk 5 above)
Comment (Unix4ever) [reply internal/pkg/encryption/keys/generated.go:41]: Yeah, but I need help with forming randomness criteria/threshold.
We can treat uuid not random if it consists of 0 more than some ratio I guess, but I'm just not sure what level of check will be enough.
I was trying to google any libraries or methods which do that, but I couldn't find anything. Maybe I was just searching for the wrong keywords.

File: internal/pkg/mount/system.go (hunk 7 above)
Comment (Unix4ever) [reply internal/pkg/mount/system.go:143]: yep

File: pkg/machinery/config/types/v1alpha1/v1alpha1_types.go (hunk 9 above)
Comment (Unix4ever) [reply pkg/machinery/config/types/v1alpha1/v1alpha1_types.go:1109]: I was asking for a better name on the review, but haven't got any good suggestions :)

File: cmd/installer/pkg/install/manifest.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (Unix4ever) [reply cmd/installer/pkg/install/manifest.go:380]: On other hand I think this check is not going to work. I've already changed that place for the state partition encryption.
I think to drop it for now, or I can extract my change from the state partition encryption branch...

File: cmd/installer/pkg/install/manifest.go (hunk 1 above)
Comment (Unix4ever) [reply cmd/installer/pkg/install/manifest.go:549]: I think I had some idea about it, but I no longer remember :shrug: Removed

File: pkg/machinery/config/types/v1alpha1/v1alpha1_types.go (hunk 8 above)
Comment (Unix4ever) [reply pkg/machinery/config/types/v1alpha1/v1alpha1_types.go:1105]: We really need that only for tests, I would like to strongly discourage using that key type in prod.
Doesn't hurt to update it to use []byte type and decode base64 though, yeah.

File: cmd/installer/pkg/install/manifest.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (Unix4ever) [reply cmd/installer/pkg/install/manifest.go:380]: I think to add the following check into `target.SaveContents`:

```go
	sb, err := source.SuperBlock()
	if err != nil {
		return err
	}

	if sb == nil || sb.Encrypted() {
		if source.Length() > partition.StateSize { // or whatever threshold we define
			return fmt.Errorf("partition %s size %s is too big to be preserved in memory", source.Name, humanize.Bytes(source.Length()))
		}

		err = t.saveRawContents(partPath)
	} else {
		err = t.saveFilesystemContents(partPath, fileSystemType, fnmatchFilters)
	}
```
This way we may get a more universal check for that case:
- if encrypted we can only preserve raw contents.
- if partition size is too big we just bail out.
- we read partition size from gpt.

What do you think?

**Review by smira — DISMISSED** [review]

**Pull request closed without merging**